### PR TITLE
fix(jit): refresh stale arch detection and allow SM120 on CUDA 12.8

### LINKED ISSUE #3675: error:  FlashInfer requires GPUs with sm75 or higher
I used a env with
5090
cuda 12.8

and got error

(StageEngineCoreProc pid=8146)   File "/root/autodl-tmp/omni/lib/python3.12/site-packages/flashinfer/sampling.py", line 1820, in top_k_mask_logits

(StageEngineCoreProc pid=8146)     return get_sampling_module().top_k_mask_logits(

(StageEngineCoreProc pid=8146)            ^^^^^^^^^^^^^^^^^^^^^

(StageEngineCoreProc pid=8146)   File "/root/autodl-tmp/omni/lib/python3.12/site-packages/flashinfer/sampling.py", line 68, in get_sampling_module

(StageEngineCoreProc pid=8146)     module = gen_sampling_module().build_and_load()

(StageEngineCoreProc pid=8146)              ^^^^^^^^^^^^^^^^^^^^^

(StageEngineCoreProc pid=8146)   File "/root/autodl-tmp/omni/lib/python3.12/site-packages/flashinfer/jit/sampling.py", line 22, in gen_sampling_module

(StageEngineCoreProc pid=8146)     return gen_jit_spec(

(StageEngineCoreProc pid=8146)            ^^^^^^^^^^^^^

(StageEngineCoreProc pid=8146)   File "/root/autodl-tmp/omni/lib/python3.12/site-packages/flashinfer/jit/core.py", line 413, in gen_jit_spec

(StageEngineCoreProc pid=8146)     check_cuda_arch()

(StageEngineCoreProc pid=8146)   File "/root/autodl-tmp/omni/lib/python3.12/site-packages/flashinfer/jit/core.py", line 108, in check_cuda_arch

(StageEngineCoreProc pid=8146)     raise RuntimeError("FlashInfer requires GPUs with sm75 or higher")

### PR DESCRIPTION
**Pull request by zhougit86**
## 📌 Description

This PR fixes two JIT architecture-detection issues on consumer Blackwell (`SM120`) environments.

- Refresh `current_compilation_context` before `check_cuda_arch()` so JIT does not rely on stale import-time arch state.
- Allow `SM120` on CUDA 12.8 in `CompilationContext._normalize_cuda_arch()`, while keeping `SM121+` gated on CUDA 12.9.
- Add regression tests for stale context refresh and `SM120` / `SM121` normalization behavior.

This is needed because PyTorch could already detect an RTX 5090 (`SM120`), but FlashInfer JIT could still fail with a misleading `sm75` error, and later with an overly strict `SM12.x requires CUDA >= 12.9` check.

## 🔍 Related Issues

- Related to SM120 / consumer Blackwell JIT detection issues
- Reproduced on RTX 5090 (`SM120`) with a CUDA 12.8 toolkit environment

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [ ] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [ ] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see `https://pre-commit.com/` .

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

Validated with:

```bash
python -m pytest tests/utils/test_jit_arch_detection.py -q -s
```

Result:

```text
5 passed in 0.03s
```

## Reviewer Notes

Main things to review:

- whether refreshing the shared JIT compilation context in `check_cuda_arch()` is the right place to fix the stale-state issue
- whether allowing `SM120` on CUDA 12.8 in `_normalize_cuda_arch()` is the right normalization policy

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved handling and reporting when CUDA device capability detection fails, including more detailed CUDA availability information.
  * Refreshed CUDA architecture detection to prevent stale results and improved validation errors by listing detected architectures.
  * Updated CUDA 12.x SM architecture normalization with explicit CUDA 12.8 and 12.9+ behavior and stricter requirements for SM 12.1+.

* **Tests**
  * Added coverage for CUDA architecture detection, refresh behavior, normalization edge cases, and error messaging.

### CODE REVIEW

**Comment by willayy** [comment]
@sricketts @dhiraj113 @aleozlx  can anyone get around to reviewing this?

**Comment by melcheikh** [comment]
Verified this PR on affected hardware — it fixes the failure end to end.

**Environment:** RTX 5090 Laptop GPU (SM120), flashinfer 0.6.13, torch 2.11.0+cu130, system nvcc 12.8 (`get_cuda_version()` prefers nvcc over `torch.version.cuda`, so this hits the `< 12.9` path).

**Before (upstream 0.6.13):** `check_cuda_arch()` raises the misleading `FlashInfer requires GPUs with sm75 or higher` — the real error (`SM 12.x requires CUDA >= 12.9`) is swallowed into a warning by `CompilationContext.__init__`, leaving `TARGET_CUDA_ARCHS` empty. This reproduces with the JIT cache both present and cleared.

**With this PR's diff applied onto 0.6.13** (fresh JIT cache):
- `check_cuda_arch()` passes; `TARGET_CUDA_ARCHS={(12, '0a')}`
- `sampling.top_k_top_p_sampling_from_logits` JIT-compiles with nvcc 12.8 (~40 s) and samples correctly on repeated calls, including a batch 64 × vocab 32k sanity check (all ids in range)
- the 5 new unit tests pass
- end-to-end vLLM generate on the same machine goes from engine death at startup (vllm-project/vllm#42393) to working FlashInfer top-k/top-p sampling

vllm-project/vllm#48956 adds a graceful fallback on the vLLM side for setups where FlashInfer can't target the GPU; this PR fixes the actual root cause for SM120 + CUDA 12.8. Would be great to see it land.

Minor note (non-blocking): the `except Exception` in `CompilationContext.__init__` still downgrades normalization errors to a warning — e.g. SM 12.1 + CUDA 12.8 would still surface as the sm75 message, now at least with `Detected TARGET_CUDA_ARCHS=[]` attached. Might be worth surfacing the stored reason in `check_cuda_arch()` in a follow-up.

File: flashinfer/jit/core.py [hunk 1]
```diff
@@ -93,10 +93,24 @@ def _print_once(self, log_method, msg: str, *args: Hashable) -> None:
...
+def refresh_current_compilation_context() -> CompilationContext:
+    """Refresh the cached compilation context in place.
+
+    Many JIT modules import ``current_compilation_context`` directly from this
+    module, so mutating the shared object keeps those aliases in sync while
+    still allowing us to re-probe the currently visible CUDA devices.
+    """
+
+    refreshed = CompilationContext()
+    current_compilation_context.TARGET_CUDA_ARCHS = refreshed.TARGET_CUDA_ARCHS
```
[the remark this answers is not in the document]
Comment (bkryu) [reply flashinfer/jit/core.py]: Can we address this comment about refresh in place? It seems valid to me.

File: flashinfer/jit/core.py [hunk 2]
```diff
@@ -93,10 +93,24 @@ def _print_once(self, log_method, msg: str, *args: Hashable) -> None:
...
+    return current_compilation_context
+
+
 def check_cuda_arch():
+    compilation_context = refresh_current_compilation_context()
```
Comment (bkryu) [inline flashinfer/jit/core.py:112]: Could we refresh at the point where architecture flags are requested instead of (only) in `check_cuda_arch()`? Many generators call `get_nvcc_flags_list()` before `gen_jit_spec()` (e.g. `gen_mm_bf16_cublaslt_module()` in flashinfer/jit/gemm/core.py), The default-flags path also goes through this method inside `cpp_ext.py` at build time. On a cold start with a stale context, those raise "No  supported CUDA architectures found" in the generator and never reach the refresh in `check_cuda_arch()`, so the failure this PR fixes still reproduces  through that path.

`refresh()` should mutate the existing set with `clear()`/`update()` to preserve aliases.

**Review by bkryu — COMMENTED** [review]
Thanks @zhougit86, for the PR. SM120 should indeed be enabled on CUDA 12.8. Left a few comments about where we should place things

File: flashinfer/jit/core.py (hunk 2 above)
Comment (zhougit86) [reply flashinfer/jit/core.py:112]: @bkryu Thanks. I agree on both parts:

1. `refresh()` should mutate `TARGET_CUDA_ARCHS` in place to preserve aliases.
2. Longer term, the cleaner fix is to avoid relying on the shared `current_compilation_context` in generator paths and construct a fresh `CompilationContext()` where flags are requested.

I’d like to keep this PR scoped to the SM120 fix and the current stale-state path, and follow up with a separate PR for the broader cleanup.

And do you mind share me the roadmap of this flashinfer project. I have been reading the cutelass related code in this project. And wish I can do some CUDA development. Thanks

File: flashinfer/jit/core.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhougit86) [reply flashinfer/jit/core.py]: yes, this makes sense. And I have fixed it.

**Comment by samelliott89** [comment]
Another SM120 confirmation, with a datapoint for the threshold choice: a **CUDA 12.6** toolkit, which is below this PR's 12.8 floor.

**Environment:** RTX PRO 6000 Blackwell Max-Q (SM120, cc 12.0), driver 610.43.02, flashinfer 0.6.17, torch 2.11.0+cu130, system nvcc 12.6.85.

Still reproduces on 0.6.17, via the batch-prefill JIT path during CUDA graph capture:

```
Failed to get device capability: SM 12.x requires CUDA >= 12.9.
...
  File ".../flashinfer/jit/attention/modules.py", line 1711, in gen_customize_batch_prefill_module
  File ".../flashinfer/jit/core.py", line 524, in gen_jit_spec
    check_cuda_arch()
RuntimeError: FlashInfer requires GPUs with sm75 or higher
```

**The 12.8 floor in this PR looks correct.** nvcc 12.6 rejects both candidate targets outright, so there is nothing to loosen further:

```
$ nvcc -gencode=arch=compute_120a,code=sm_120a -c t.cu
nvcc fatal : Unsupported gpu architecture 'compute_120a'
$ nvcc -gencode=arch=compute_120f,code=sm_120f -c t.cu
nvcc fatal : Unsupported gpu architecture 'compute_120f'
```

What this PR would have changed for me is the diagnostic. `SM 120 requires CUDA >= 12.8` names the problem and the fix; `requires GPUs with sm75 or higher` on a cc-12.0 card sends you off checking the GPU, the driver, and `torch.cuda` — all of which were fine (`torch.cuda.get_device_capability(0)` returned `(12, 0)` the whole time). That improvement is worth landing on its own, independent of the 12.8 normalization policy.

One more wrinkle worth flagging, since it decides *which* CUDA version the gate reads. `get_cuda_version()` prefers nvcc and only falls back to `torch.version.cuda`, so on this box the two disagree and the older one wins:

```
torch.version.cuda      = 13.0
flashinfer picks        = 12.6   # from /usr/local/cuda/bin/nvcc
```

The torch install is `+cu130` and its CUDA 13 headers and libs are already sitting in `site-packages/nvidia/cu13` — only nvcc itself is absent from that prefix. So the "wrong" toolkit is the one on PATH, while a suitable one ships with the environment. Not arguing nvcc-first is the wrong precedence, but it's worth a line in the error message or docs: pointing `CUDA_HOME` at the torch-provided prefix is a fix many users already have most of the pieces for.

Two smaller notes for anyone landing here:

- **`FLASHINFER_CUDA_ARCH_LIST=12.0f` is not a workaround.** A suffixed value skips `_normalize_cuda_arch` and passes `check_cuda_arch()`, but the build then dies at `nvcc fatal : Unsupported gpu architecture 'compute_120f'`. It converts a misleading early error into a later one.
- **A prebuilt cache doesn't avoid this.** `flashinfer-jit-cache` 0.6.17+cu130 ships sm120 kernels (`nvfp4_attention_sm120`, `fp4_gemm_cutlass_sm120`, …), but `gen_batch_prefill_module` still went to JIT, so nvcc remains on the critical path.

**Fix that worked**, without touching the system toolkit: install the `nvidia-cuda-nvcc` wheel (13.3.73) into the venv and point `CUDA_HOME` at its `nvidia/cu13` prefix, which already carries matching CUDA 13 headers and libs. Detection then yields `{(12, '0f')}`, the prefill module JIT-compiles, CUDA graph capture completes, and the model serves — verified end to end with a stripped environment (`env -i`), so nothing depended on my shell. Note `nvidia-cuda-nvcc-cu13` is deprecated and fails to build — the unsuffixed name is the one to use.

Happy to test this PR on SM120 + CUDA 12.6 if that's useful for the error-path change.

**Comment by manojhf** [comment]
Can we merge this PR? What is stopping it?

**Comment by melcheikh** [comment]
Second hardware datapoint, and a note on how vLLM is handling this on its side.

**RTX PRO 6000 Blackwell Server Edition (SM120, 96 GB), GCP g4-standard-48, driver 580.173.02, flashinfer-python 0.6.18, torch 2.13.0+cu132.** The VM ships CUDA 12.9; I installed a CUDA 12.8.1 toolkit side by side and switched `CUDA_HOME`/`PATH` to it, which is exactly the `get_cuda_version() < 12.9` path this PR changes. On 0.6.18 with nvcc 12.8 every nvcc-JIT entry point still fails at first use: `gen_cutlass_fused_moe_sm120_module` and the FP8/FP4 GEMM modules raise `No supported CUDA architectures found for major versions [12]`, generic specs raise `FlashInfer requires GPUs with sm75 or higher`. With nvcc 12.9 on the same box the same modules compile and run (FP8 GEMM, CUTLASS fused-MoE, CUTLASS NVFP4 GEMM all produced correct generations), so nothing beyond the version gate is missing for SM120. That matches what I saw on the RTX 5090 in July.

On the vLLM side I opened vllm-project/vllm#55651: at kernel-selection time it checks whether the current GPU's arch is in `current_compilation_context.TARGET_CUDA_ARCHS` and skips FlashInfer-backed kernels when it is not, re-deriving the swallowed `SM 12.x requires CUDA >= 12.9` message for the log. It reads FlashInfer's state rather than a version number, so once this PR lands the check passes on 12.8 and nothing changes there. It is a fallback, not a substitute for this fix: SM121 stays on 12.9 as this PR keeps it, and the CUDA 12.6 report above stays below the floor either way.

If it helps move this along: @zhougit86, are you still working on the placement comments from @bkryu? If not, I can put up a follow-up PR that carries your commits with those comments addressed, keeping your authorship.

**Comment by bkryu** [comment]
@flashinfer-bot run

**Comment by bkryu** [comment]
Thanks @zhougit86  and @melcheikh for your patience on the PR. The PR looks good to me and we need the fix.

Triggered the CI. Had to merge main to get stuck pre-commit test job unstuck. Once the CI passes, should be good to go

**Comment by bkryu** [comment]
@flashinfer-bot run

**Comment by bkryu** [comment]
@flashinfer-bot run

**Comment by bkryu** [comment]
@flashinfer-bot run

**Review by bkryu — APPROVED** [review]
